File: ai/spring/news/src/main/java/kr/yeotaeho/api/service/NewsService.java
package kr.yeotaeho.api.service;

import kr.yeotaeho.api.config.RssUrlMapper;
import kr.yeotaeho.api.dto.NaverNewsResponse;
import kr.yeotaeho.api.dto.NewsArticle;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.stream.Collectors;

@Slf4j
@Service
public class NewsService {

    private final RestTemplate restTemplate;
    private final RssService rssService;
    private final RssUrlMapper rssUrlMapper;
    private final ExecutorService rssExecutor;
    
    public NewsService(
            RestTemplate restTemplate,
            RssService rssService,
            RssUrlMapper rssUrlMapper,
            @Qualifier("rssExecutor") ExecutorService rssExecutor
    ) {
        this.restTemplate = restTemplate;
        this.rssService = rssService;
        this.rssUrlMapper = rssUrlMapper;
        this.rssExecutor = rssExecutor;
    }

    @Value("${naver.client-id}")
    private String naverClientId;

    @Value("${naver.client-secret}")
    private String naverClientSecret;

    private static final String NAVER_NEWS_API_URL = "https://openapi.naver.com/v1/search/news.json";

    /**
     * 뉴스 검색 (RSS 우선, 필요시 네이버 API 폴백)
     * 
     * @param query 검색어 또는 카테고리명
     * @param display 표시할 결과 수 (기본값: 20)
     * @param start 시작 위치 (기본값: 1)
     * @return 뉴스 기사 목록
     */
    public List<NewsArticle> searchNews(String query, Integer display, Integer start) {
        // 1. query가 정의된 카테고리명인지 확인
        if (rssUrlMapper.isCategory(query)) {
            log.info("카테고리로 인식: query={}, RSS 피드 사용", query);
            return searchNewsFromRss(query, display, start);
        } else {
            log.info("검색어로 인식: query={}, 네이버 API 사용", query);
            // 자유 검색어 → 네이버 API 호출 (제한적)
            return searchNewsFromNaverApi(query, display, start);
        }
    }
    
    /**
     * RSS 피드를 통한 뉴스 검색 (비동기 병렬 처리)
     */
    private List<NewsArticle> searchNewsFromRss(String category, Integer display, Integer start) {
        try {
            // 1. 카테고리별 RSS URL 목록 조회
            List<String> rssUrls = rssUrlMapper.getRssUrlsByCategory(category);
            
            if (rssUrls.isEmpty()) {
                log.warn("카테고리에 해당하는 RSS URL이 없습니다: category={}", category);
                return List.of();
            }
            
            log.info("RSS 피드 수집 시작: category={}, RSS URL 개수={}", category, rssUrls.size());
            
            // 2. 비동기 병렬 처리로 여러 RSS 피드 수집
            List<NewsArticle> allArticles = fetchMultipleRssFeeds(rssUrls);
            
            log.info("RSS 피드 수집 완료: category={}, 총 기사 수={}", category, allArticles.size());
            
            // 3. 중복 제거 (제목 기준) - LinkedHashMap으로 순서 보장
            List<NewsArticle> uniqueArticles = allArticles.stream()
                    .collect(Collectors.toMap(
                            NewsArticle::getTitle,
                            article -> article,
                            (existing, replacement) -> existing,
                            LinkedHashMap::new
                    ))
                    .values()
                    .stream()
                    .collect(Collectors.toList());
            
            log.info("중복 제거 후 기사 수: category={}, 기사 수={}", category, uniqueArticles.size());
            
            // 4. 날짜순 정렬 (이미 RssService에서 정렬되어 있을 수 있지만 재정렬)
            uniqueArticles.sort((a, b) -> {
                try {
                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy.MM.dd");
                    java.time.LocalDate dateA = java.time.LocalDate.parse(a.getDate(), formatter);
                    java.time.LocalDate dateB = java.time.LocalDate.parse(b.getDate(), formatter);
                    return dateB.compareTo(dateA); // 내림차순 (최신이 먼저)
                } catch (Exception e) {
                    log.warn("날짜 정렬 실패: dateA={}, dateB={}", a.getDate(), b.getDate());
                    return 0;
                }
            });
            
            // 5. 페이징 처리
            return applyPaging(uniqueArticles, display, start);
            
        } catch (Exception e) {
            log.error("RSS 뉴스 검색 실패: category={}, error={}", category, e.getMessage(), e);
            return List.of();
        }
    }
    
    /**
     * 여러 RSS 피드를 비동기 병렬로 수집
     */
    private List<NewsArticle> fetchMultipleRssFeeds(List<String> rssUrls) {
        // 1. 각 RSS URL에 대해 CompletableFuture 생성
        List<CompletableFuture<List<NewsArticle>>> futures = rssUrls.stream()
                .map(url -> CompletableFuture.supplyAsync(() -> {
                    try {
                        return rssService.fetchNewsFromRss(url);
                    } catch (Exception e) {
                        log.error("RSS 피드 수집 실패: URL={}, error={}", url, e.getMessage());
                        return List.<NewsArticle>of(); // 빈 리스트 반환
                    }
                }, rssExecutor))
                .collect(Collectors.toList());
        
        // 2. 모든 Future가 완료될 때까지 대기
        CompletableFuture<Void> allFutures = CompletableFuture.allOf(
                futures.toArray(new CompletableFuture[0])
        );
        
        // 3. 모든 결과 수집 및 병합
        return allFutures.thenApply(v -> 
                futures.stream()
                        .map(CompletableFuture::join)
                        .flatMap(List::stream)
                        .collect(Collectors.toList())
        ).join();
    }
    
    /**
     * 메모리 페이징 처리
     */
    private List<NewsArticle> applyPaging(
            List<NewsArticle> articles, 
            Integer display, 
            Integer start
    ) {
        // 기본값 설정
        if (display == null || display <= 0) {
            display = 20;
        }
        if (start == null || start < 1) {
            start = 1;
        }
        
        // start 파라미터로 offset 적용
        int offset = start - 1;
        int fromIndex = Math.min(offset, articles.size());
        
        // display 파라미터로 limit 적용
        int toIndex = Math.min(fromIndex + display, articles.size());
        
        // 서브리스트 반환
        if (fromIndex >= toIndex) {
            return List.of();
        }
        
        return articles.subList(fromIndex, toIndex);
    }
    
    /**
     * 네이버 API를 통한 뉴스 검색 (폴백)
     */
    private List<NewsArticle> searchNewsFromNaverApi(String query, Integer display, Integer start) {
        try {
            // 기본값 설정
            if (display == null || display <= 0) {
                display = 20;
            }
            if (start == null || start < 1) {
                start = 1;
            }

            // URL 생성
            UriComponentsBuilder builder = UriComponentsBuilder.fromHttpUrl(NAVER_NEWS_API_URL)
                    .queryParam("query", query)
                    .queryParam("display", display)
                    .queryParam("start", start)
                    .queryParam("sort", "date"); // 날짜순 정렬

            // 헤더 설정
            HttpHeaders headers = new HttpHeaders();
            headers.set("X-Naver-Client-Id", naverClientId);
            headers.set("X-Naver-Client-Secret", naverClientSecret);
            HttpEntity<String> entity = new HttpEntity<>(headers);

            // API 호출
            ResponseEntity<NaverNewsResponse> response = restTemplate.exchange(
                    builder.toUriString(),
                    HttpMethod.GET,
                    entity,
                    NaverNewsResponse.class
            );

            NaverNewsResponse newsResponse = response.getBody();
            if (newsResponse == null || newsResponse.getItems() == null) {
                log.warn("네이버 뉴스 API 응답이 비어있습니다.");
                return List.of();
            }

            log.info("네이버 뉴스 API 응답: 총 {}개 기사", newsResponse.getItems().size());

            // 원본 pubDate를 사용하여 날짜 기준 내림차순 정렬 (최신 기사 우선)
            DateTimeFormatter rfc822Formatter = DateTimeFormatter.RFC_1123_DATE_TIME;
            //ZonedDateTime now = ZonedDateTime.now();
            //ZonedDateTime sevenDaysAgo = now.minusDays(7); // 최근 7일 이내 기사만

            // 날짜 기준 내림차순 정렬 및 최근 7일 이내 기사만 필터링
            List<NaverNewsResponse.NaverNewsItem> sortedItems = newsResponse.getItems().stream()
                    //.filter(item -> {
                    //    try {
                    //        ZonedDateTime pubDate = ZonedDateTime.parse(item.getPubDate(), rfc822Formatter);
                    //        // 최근 7일 이내 기사만 포함
                    //        return !pubDate.isBefore(sevenDaysAgo);
                    //    } catch (Exception e) {
                    //        log.warn("날짜 필터링 실패: pubDate={}, error={}", item.getPubDate(), e.getMessage());
                    //        return false; // 날짜 파싱 실패 시 제외
                    //    }
                    //})
                    .sorted((a, b) -> {
                        // 원본 pubDate를 사용하여 정렬 (최신 기사가 먼저)
                        try {
                            ZonedDateTime dateA = ZonedDateTime.parse(a.getPubDate(), rfc822Formatter);
                            ZonedDateTime dateB = ZonedDateTime.parse(b.getPubDate(), rfc822Formatter);
                            return dateB.compareTo(dateA); // 내림차순 (최신이 먼저)
                        } catch (Exception e) {
                            log.warn("날짜 정렬 실패: pubDate={}, error={}", a.getPubDate(), e.getMessage());
                            return 0;
                        }
                    })
                    .collect(Collectors.toList());

            log.info("정렬 및 필터링 후 기사 수: {}개 (최근 7일 이내)", sortedItems.size());

            // DTO 변환
            List<NewsArticle> articles = sortedItems.stream()
                    .map(item -> convertToNewsArticle(item, query))
                    .collect(Collectors.toList());

            log.info("최종 변환된 기사 수: {}개", articles.size());
            return articles;

        } catch (Exception e) {
            log.error("네이버 뉴스 API 호출 실패: {}", e.getMessage(), e);
            return List.of();
        }
    }

    /**
     * 네이버 뉴스 아이템을 NewsArticle로 변환
     */
    private NewsArticle convertToNewsArticle(NaverNewsResponse.NaverNewsItem item, String query) {
        // HTML 태그 제거
        String cleanTitle = item.getTitle()
                .replaceAll("<[^>]*>", "")
                .replaceAll("&quot;", "\"")
                .replaceAll("&amp;", "&")
                .replaceAll("&lt;", "<")
                .replaceAll("&gt;", ">");

        String cleanDescription = item.getDescription()
                .replaceAll("<[^>]*>", "")
                .replaceAll("&quot;", "\"")
                .replaceAll("&amp;", "&")
                .replaceAll("&lt;", "<")
                .replaceAll("&gt;", ">");

        // 날짜 포맷팅 (네이버 API는 RFC 822 형식: "Wed, 13 Dec 2025 10:30:00 +0900")
        String formattedDate = formatDate(item.getPubDate());

        // 이미지 추출 (설명에서 이미지 URL이 있는 경우)
        String imageUrl = extractImageUrl(item.getDescription());

        return NewsArticle.builder()
                .type(query != null ? query : "뉴스")
                .title(cleanTitle)
                .date(formattedDate)
                .image(imageUrl)
                .link(item.getLink())
                .description(cleanDescription)
                .build();
    }

    /**
     * 날짜 포맷팅 (RFC 822 -> YYYY.MM.DD)
     */
    private String formatDate(String pubDate) {
        try {
            // RFC 822 형식 파싱 (예: "Wed, 13 Dec 2025 10:30:00 +0900")
            DateTimeFormatter rfc822Formatter = DateTimeFormatter.RFC_1123_DATE_TIME;
            ZonedDateTime zonedDateTime = ZonedDateTime.parse(pubDate, rfc822Formatter);
            return zonedDateTime.format(DateTimeFormatter.ofPattern("yyyy.MM.dd"));
        } catch (Exception e) {
            log.warn("날짜 파싱 실패: {}, 현재 날짜 반환", pubDate);
            // 날짜 파싱 실패 시 현재 날짜 반환
            return LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy.MM.dd"));
        }
    }

    /**
     * 설명에서 이미지 URL 추출
     */
    private String extractImageUrl(String description) {
        // 간단한 이미지 URL 추출 (실제로는 더 복잡한 로직이 필요할 수 있음)
        if (description == null) {
            return "https://placehold.co/400x250/000000/FFFFFF?text=NEWS";
        }
        
        // img 태그에서 src 추출 시도
        String imgPattern = "<img[^>]+src\\s*=\\s*['\"]([^'\"]+)['\"]";
        java.util.regex.Pattern pattern = java.util.regex.Pattern.compile(imgPattern);
        java.util.regex.Matcher matcher = pattern.matcher(description);
        
        if (matcher.find()) {
            return matcher.group(1);
        }
        
        // 이미지 URL이 없으면 기본 이미지
        return "https://placehold.co/400x250/000000/FFFFFF?text=NEWS";
    }
}

File: backend/domain/news/service/news_service.py
import asyncio
import httpx
from typing import List, Optional
import logging
from bs4 import BeautifulSoup
from datetime import datetime
from ..model.news_article import NewsArticle
from ..config.rss_url_mapper import RssUrlMapper
from .rss_service import RssService

logger = logging.getLogger(__name__)


class NewsService:
    """뉴스 서비스 (RSS 및 네이버 API 통합)"""
    
    def __init__(self, naver_client_id: str, naver_client_secret: str):
        """
        NewsService 초기화
        
        Args:
            naver_client_id: 네이버 API 클라이언트 ID
            naver_client_secret: 네이버 API 클라이언트 시크릿
        """
        self.naver_client_id = naver_client_id
        self.naver_client_secret = naver_client_secret
        self.rss_service = RssService()
        self.rss_url_mapper = RssUrlMapper()
        self.naver_api_url = "https://openapi.naver.com/v1/search/news.json"
    
    async def search_news(
        self, 
        query: str, 
        display: Optional[int] = None, 
        start: Optional[int] = None
    ) -> List[NewsArticle]:
        """
        뉴스 검색 (RSS 우선, 필요시 네이버 API)
        
        Args:
            query: 검색어 또는 카테고리명
            display: 표시할 결과 수
            start: 시작 위치
            
        Returns:
            뉴스 기사 목록
        """
        if self.rss_url_mapper.is_category(query):
            logger.info(f"카테고리로 인식: query={query}, RSS 피드 사용")
            return await self._search_news_from_rss(query, display, start)
        else:
            logger.info(f"검색어로 인식: query={query}, 네이버 API 사용")
            return await self._search_news_from_naver_api(query, display, start)
    
    async def _search_news_from_rss(
        self, 
        category: str, 
        display: Optional[int], 
        start: Optional[int]
    ) -> List[NewsArticle]:
        """RSS 피드를 통한 뉴스 검색 (비동기 병렬 처리)"""
        try:
            rss_urls = self.rss_url_mapper.get_rss_urls_by_category(category)
            
            if not rss_urls:
                logger.warning(f"카테고리에 해당하는 RSS URL이 없습니다: category={category}")
                return []
            
            logger.info(f"RSS 피드 수집 시작: category={category}, RSS URL 개수={len(rss_urls)}")
            
            # 비동기 병렬 처리로 여러 RSS 피드 수집
            all_articles = await self._fetch_multiple_rss_feeds(rss_urls)
            
            logger.info(f"RSS 피드 수집 완료: category={category}, 총 기사 수={len(all_articles)}")
            
            # 중복 제거 (제목 기준)
            unique_articles = self._remove_duplicates(all_articles)
            
            logger.info(f"중복 제거 후 기사 수: category={category}, 기사 수={len(unique_articles)}")
            
            # 페이징 처리
            return self._apply_paging(unique_articles, display, start)
            
        except Exception as e:
            logger.error(f"RSS 뉴스 검색 실패: category={category}, error={e}", exc_info=True)
            return []
    
    async def _fetch_multiple_rss_feeds(self, rss_urls: List[str]) -> List[NewsArticle]:
        """여러 RSS 피드를 비동기 병렬로 수집"""
        # 동기 함수를 비동기로 실행
        loop = asyncio.get_event_loop()
        tasks = [
            loop.run_in_executor(None, self.rss_service.fetch_news_from_rss, url)
            for url in rss_urls
        ]
        
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        all_articles = []
        for i, result in enumerate(results):
            if isinstance(result, Exception):
                logger.error(f"RSS 피드 수집 실패: URL={rss_urls[i]}, error={result}")
                continue
            all_articles.extend(result)
        
        return all_articles
    
    def _remove_duplicates(self, articles: List[NewsArticle]) -> List[NewsArticle]:
        """중복 제거 (제목 기준)"""
        seen = set()
        unique = []
        for article in articles:
            if article.title not in seen:
                seen.add(article.title)
                unique.append(article)
        return unique
    
    def _apply_paging(
        self, 
        articles: List[NewsArticle], 
        display: Optional[int], 
        start: Optional[int]
    ) -> List[NewsArticle]:
        """페이징 처리"""
        display = display or 20
        start = start or 1
        
        offset = start - 1
        from_index = min(offset, len(articles))
        to_index = min(from_index + display, len(articles))
        
        if from_index >= to_index:
            return []
        
        return articles[from_index:to_index]
    
    async def _search_news_from_naver_api(
        self, 
        query: str, 
        display: Optional[int], 
        start: Optional[int]
    ) -> List[NewsArticle]:
        """네이버 API를 통한 뉴스 검색"""
        try:
            display = display or 20
            start = start or 1
            
            async with httpx.AsyncClient(timeout=30.0) as client:
                response = await client.get(
                    self.naver_api_url,
                    params={
                        "query": query,
                        "display": display,
                        "start": start,
                        "sort": "date"
                    },
                    headers={
                        "X-Naver-Client-Id": self.naver_client_id,
                        "X-Naver-Client-Secret": self.naver_client_secret
                    }
                )
                response.raise_for_status()
                data = response.json()
                
                articles = []
                items = data.get("items", [])
                
                # 날짜순 정렬 (최신이 먼저)
                items.sort(key=lambda x: self._parse_naver_date(x.get("pubDate", "")), reverse=True)
                
                for item in items:
                    try:
                        article = self._convert_naver_item_to_article(item, query)
                        articles.append(article)
                    except Exception as e:
                        logger.warning(f"네이버 기사 변환 실패: {e}")
                        continue
                
                logger.info(f"네이버 뉴스 API 응답: 총 {len(articles)}개 기사")
                return articles
                
        except Exception as e:
            logger.error(f"네이버 뉴스 API 호출 실패: {e}", exc_info=True)
            return []
    
    def _convert_naver_item_to_article(self, item: dict, query: str) -> NewsArticle:
        """네이버 뉴스 아이템을 NewsArticle로 변환"""
        # HTML 태그 제거
        title = self._clean_html(item.get("title", ""))
        description = self._clean_html(item.get("description", ""))
        
        # 날짜 포맷팅 (RFC 822 -> yyyy.MM.dd)
        date = self._format_naver_date(item.get("pubDate", ""))
        
        # 이미지 추출
        image_url = self._extract_image_from_description(item.get("description", ""))
        
        return NewsArticle(
            type=query or "뉴스",
            title=title,
            date=date,
            image=image_url,
            link=item.get("link", ""),
            description=description
        )
    
    def _clean_html(self, html: str) -> str:
        """HTML 태그 제거"""
        if not html:
            return ""
        
        try:
            soup = BeautifulSoup(html, "html.parser")
            text = soup.get_text()
            # HTML 엔티티 디코딩
            text = text.replace("&quot;", "\"").replace("&amp;", "&")
            text = text.replace("&lt;", "<").replace("&gt;", ">")
            return text.strip()
        except Exception as e:
            logger.warning(f"HTML 정리 실패: {e}")
            return html
    
    def _format_naver_date(self, pub_date: str) -> str:
        """네이버 API 날짜 포맷팅 (RFC 822 -> yyyy.MM.dd)"""
        if not pub_date:
            return datetime.now().strftime("%Y.%m.%d")
        
        try:
            # RFC 822 형식 파싱 (예: "Wed, 13 Dec 2025 10:30:00 +0900")
            # Python의 email.utils.parsedate_to_datetime 사용
            from email.utils import parsedate_to_datetime
            dt = parsedate_to_datetime(pub_date)
            return dt.strftime("%Y.%m.%d")
        except Exception as e:
            logger.warning(f"날짜 포맷팅 실패: pubDate={pub_date}, {e}")
            return datetime.now().strftime("%Y.%m.%d")
    
    def _parse_naver_date(self, pub_date: str) -> datetime:
        """네이버 API 날짜를 datetime으로 파싱 (정렬용)"""
        try:
            from email.utils import parsedate_to_datetime
            return parsedate_to_datetime(pub_date)
        except:
            return datetime.min
    
    def _extract_image_from_description(self, description: str) -> str:
        """설명에서 이미지 URL 추출"""
        if not description:
            return "https://placehold.co/400x250/000000/FFFFFF?text=NEWS"
        
        try:
            soup = BeautifulSoup(description, "html.parser")
            img = soup.find("img")
            
            if img:
                # src, data-src 순서로 확인
                for attr in ["src", "data-src", "data-lazy-src"]:
                    url = img.get(attr, "")
                    if url:
                        if url.startswith("//"):
                            return "https:" + url
                        elif url.startswith("http://") or url.startswith("https://"):
                            return url
        except Exception as e:
            logger.debug(f"이미지 추출 실패: {e}")
        
        return "https://placehold.co/400x250/000000/FFFFFF?text=NEWS"

